src/bootstrap/setup.rs

```diff
@@ -71,13 +82,50 @@ impl FromStr for Profile {
 
 impl fmt::Display for Profile {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Profile::Compiler => write!(f, "compiler"),
-            Profile::Codegen => write!(f, "codegen"),
-            Profile::Library => write!(f, "library"),
-            Profile::User => write!(f, "user"),
-            Profile::Tools => write!(f, "tools"),
+        f.write_str(self.as_str())
+    }
+}
+
+impl Step for Profile {
+    type Output = ();
+    const DEFAULT: bool = true;
+
+    fn should_run(mut run: ShouldRun<'_>) -> ShouldRun<'_> {
+        for choice in Profile::all() {
+            run = run.alias(choice.as_str());
         }
+        run
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        if run.builder.config.dry_run {
+            return;
+        }
+
+        let profile: Profile = if run.paths.len() > 1 {
+            println!("\nat most one profile can be passed to setup\n");
+            crate::detail_exit(1);
+        } else if run.paths.len() == 1 {
+            let path_buf = &run.paths[0].assert_single_path().path;
+            let profile_string = t!(path_buf
+                .to_path_buf()
+                .into_os_string()
+                .into_string()
+                .map_err(|path| format!("{} is not a valid UTF8 string", path.to_string_lossy())));
+            profile_string.parse().unwrap_or_else(|err| {
+                eprintln!("error: {}", err);
+                eprintln!("help: the available profiles are:");
+                eprint!("{}", Profile::all_for_help("- "));
+                crate::detail_exit(1);
+            })
+        } else {
+            t!(crate::setup::interactive_path())
+        };
```

Comment:
I think Mark's suggestion in https://github.com/rust-lang/rust/pull/96584#issuecomment-1221323383 is to move this to flags.rs: https://github.com/rust-lang/rust/pull/96584/files#diff-f8ccbc55b1d2673954d5a025a69985cdb69ce7ceb0f6078d14593a22a1072685R603